File: frontend/app/lib/domain/entities/revenue_share_tier.dart
import 'package:equatable/equatable.dart';

/// Shopify revenue share tier for an app
/// Based on Shopify's Reduced Revenue Share Plan
enum RevenueShareTier {
  /// Default 20% revenue share (not registered for reduced plan)
  default20('DEFAULT_20', 'Default (20%)', 20.0, 'Not registered for reduced revenue share plan'),

  /// 0% revenue share on first $1M lifetime (small developer)
  smallDev0('SMALL_DEV_0', 'Small Developer (0%)', 0.0, '0% on first \$1M lifetime (under \$1M earned)'),

  /// 15% revenue share after $1M lifetime (small developer)
  smallDev15('SMALL_DEV_15', 'Small Developer (15%)', 15.0, '15% after \$1M lifetime earnings'),

  /// 15% revenue share on all revenue (large developer)
  largeDev15('LARGE_DEV_15', 'Large Developer (15%)', 15.0, '15% on all revenue (large developer)');

  final String code;
  final String displayName;
  final double revenueSharePercent;
  final String description;

  const RevenueShareTier(this.code, this.displayName, this.revenueSharePercent, this.description);

  /// Processing fee is always 2.9% regardless of tier
  static const double processingFeePercent = 2.9;

  /// Whether this is a reduced plan tier
  bool get isReducedPlan => this != default20;

  /// Parse tier from API string
  static RevenueShareTier fromCode(String? code) {
    if (code == null) return default20;
    return RevenueShareTier.values.firstWhere(
      (t) => t.code == code,
      orElse: () => default20,
    );
  }

  /// Badge color for the tier
  int get badgeColor {
    switch (this) {
      case RevenueShareTier.default20:
        return 0xFFEF4444; // Red
      case RevenueShareTier.smallDev0:
        return 0xFF22C55E; // Green
      case RevenueShareTier.smallDev15:
        return 0xFFF59E0B; // Amber
      case RevenueShareTier.largeDev15:
        return 0xFFF59E0B; // Amber
    }
  }
}

/// Fee breakdown for a transaction or gross amount
class FeeBreakdown extends Equatable {
  final int grossAmountCents;
  final int revenueShareCents;
  final int processingFeeCents;
  final int taxOnFeesCents;
  final int totalFeesCents;
  final int netAmountCents;
  final double revenueSharePercent;
  final double processingFeePercent;

  const FeeBreakdown({
    required this.grossAmountCents,
    required this.revenueShareCents,
    required this.processingFeeCents,
    required this.taxOnFeesCents,
    required this.totalFeesCents,
    required this.netAmountCents,
    required this.revenueSharePercent,
    required this.processingFeePercent,
  });

  /// Calculate fee breakdown for a given tier and gross amount
  factory FeeBreakdown.calculate({
    required int grossAmountCents,
    required RevenueShareTier tier,
    double taxRate = 0.08, // Default 8% tax on fees
  }) {
    final revenueShareCents = (grossAmountCents * tier.revenueSharePercent / 100).round();
    final processingFeeCents = (grossAmountCents * RevenueShareTier.processingFeePercent / 100).round();
    final taxOnFeesCents = ((revenueShareCents + processingFeeCents) * taxRate).round();
    final totalFeesCents = revenueShareCents + processingFeeCents + taxOnFeesCents;
    final netAmountCents = grossAmountCents - totalFeesCents;

    return FeeBreakdown(
      grossAmountCents: grossAmountCents,
      revenueShareCents: revenueShareCents,
      processingFeeCents: processingFeeCents,
      taxOnFeesCents: taxOnFeesCents,
      totalFeesCents: totalFeesCents,
      netAmountCents: netAmountCents,
      revenueSharePercent: tier.revenueSharePercent,
      processingFeePercent: RevenueShareTier.processingFeePercent,
    );
  }

  @override
  List<Object?> get props => [
        grossAmountCents,
        revenueShareCents,
        processingFeeCents,
        taxOnFeesCents,
        totalFeesCents,
        netAmountCents,
      ];
}

File: backend/internal/domain/valueobject/revenue_share_tier.go
package valueobject

// RevenueShareTier represents the Shopify revenue share tier for an app
// Based on Shopify's Reduced Revenue Share Plan:
// - DEFAULT: 20% revenue share (not registered for reduced plan)
// - SMALL_DEV_0: 0% on first $1M lifetime (registered, <$20M app / <$100M company)
// - SMALL_DEV_15: 15% after $1M lifetime (registered, <$20M app / <$100M company)
// - LARGE_DEV: 15% on all revenue (registered, ≥$20M app OR ≥$100M company)
type RevenueShareTier string

const (
	// RevenueShareTierDefault is 20% revenue share (not registered for reduced plan)
	RevenueShareTierDefault RevenueShareTier = "DEFAULT_20"

	// RevenueShareTierSmallDev0 is 0% revenue share on first $1M lifetime
	// Eligible: <$20M prior-year app revenue AND <$100M company revenue
	RevenueShareTierSmallDev0 RevenueShareTier = "SMALL_DEV_0"

	// RevenueShareTierSmallDev15 is 15% revenue share after $1M lifetime
	// Eligible: <$20M prior-year app revenue AND <$100M company revenue
	RevenueShareTierSmallDev15 RevenueShareTier = "SMALL_DEV_15"

	// RevenueShareTierLargeDev is 15% revenue share on all revenue
	// Eligible: ≥$20M prior-year app revenue OR ≥$100M company revenue
	RevenueShareTierLargeDev RevenueShareTier = "LARGE_DEV_15"
)

// Processing fee is always 2.9% regardless of tier
const ProcessingFeePercent = 2.9

func (t RevenueShareTier) String() string {
	return string(t)
}

func (t RevenueShareTier) IsValid() bool {
	switch t {
	case RevenueShareTierDefault, RevenueShareTierSmallDev0, RevenueShareTierSmallDev15, RevenueShareTierLargeDev:
		return true
	}
	return false
}

// RevenueSharePercent returns the revenue share percentage for this tier
func (t RevenueShareTier) RevenueSharePercent() float64 {
	switch t {
	case RevenueShareTierDefault:
		return 20.0
	case RevenueShareTierSmallDev0:
		return 0.0
	case RevenueShareTierSmallDev15, RevenueShareTierLargeDev:
		return 15.0
	default:
		return 20.0 // Default to 20% if unknown
	}
}

// DisplayName returns a human-readable name for the tier
func (t RevenueShareTier) DisplayName() string {
	switch t {
	case RevenueShareTierDefault:
		return "Default (20%)"
	case RevenueShareTierSmallDev0:
		return "Small Developer (0%)"
	case RevenueShareTierSmallDev15:
		return "Small Developer (15%)"
	case RevenueShareTierLargeDev:
		return "Large Developer (15%)"
	default:
		return "Unknown"
	}
}

// Description returns a description of the tier eligibility
func (t RevenueShareTier) Description() string {
	switch t {
	case RevenueShareTierDefault:
		return "Not registered for reduced revenue share plan"
	case RevenueShareTierSmallDev0:
		return "0% on first $1M lifetime (under $1M earned)"
	case RevenueShareTierSmallDev15:
		return "15% after $1M lifetime earnings"
	case RevenueShareTierLargeDev:
		return "15% on all revenue (large developer)"
	default:
		return "Unknown tier"
	}
}

// IsReducedPlan returns true if registered for reduced revenue share plan
func (t RevenueShareTier) IsReducedPlan() bool {
	return t != RevenueShareTierDefault
}

// CalculateRevenueShareCents calculates the revenue share fee in cents
func (t RevenueShareTier) CalculateRevenueShareCents(grossAmountCents int64) int64 {
	percent := t.RevenueSharePercent()
	return int64(float64(grossAmountCents) * percent / 100.0)
}

// CalculateProcessingFeeCents calculates the 2.9% processing fee in cents
func CalculateProcessingFeeCents(grossAmountCents int64) int64 {
	return int64(float64(grossAmountCents) * ProcessingFeePercent / 100.0)
}

// FeeBreakdown contains the calculated fee breakdown for a transaction
type FeeBreakdown struct {
	GrossAmountCents     int64
	RevenueShareCents    int64
	ProcessingFeeCents   int64
	TaxOnFeesCents       int64 // Estimated, actual from Shopify
	TotalFeesCents       int64
	NetAmountCents       int64
	RevenueSharePercent  float64
	ProcessingFeePercent float64
}

// CalculateFeeBreakdown calculates the complete fee breakdown for a gross amount
// taxRate is the estimated tax rate on fees (e.g., 0.08 for 8%)
func (t RevenueShareTier) CalculateFeeBreakdown(grossAmountCents int64, taxRate float64) FeeBreakdown {
	revenueShare := t.CalculateRevenueShareCents(grossAmountCents)
	processingFee := CalculateProcessingFeeCents(grossAmountCents)
	taxOnFees := int64(float64(revenueShare+processingFee) * taxRate)
	totalFees := revenueShare + processingFee + taxOnFees
	netAmount := grossAmountCents - totalFees

	return FeeBreakdown{
		GrossAmountCents:     grossAmountCents,
		RevenueShareCents:    revenueShare,
		ProcessingFeeCents:   processingFee,
		TaxOnFeesCents:       taxOnFees,
		TotalFeesCents:       totalFees,
		NetAmountCents:       netAmount,
		RevenueSharePercent:  t.RevenueSharePercent(),
		ProcessingFeePercent: ProcessingFeePercent,
	}
}

// ParseRevenueShareTier parses a string into a RevenueShareTier
func ParseRevenueShareTier(s string) RevenueShareTier {
	tier := RevenueShareTier(s)
	if tier.IsValid() {
		return tier
	}
	return RevenueShareTierSmallDev0 // Default to 0% for most indie devs
}
